[DirectX][ShaderFlags] Add analysis for `WaveOps` flag

inbelic opened the pull request:
- Check each call instruction for a `WaveOp` intrinsic and set the `WaveOps` flag if this is true for any intrinsic, Done in DXILShaderFlags.cpp

Resolves #114565

inbelic commented on llvm/test/CodeGen/DirectX/ShaderFlags/iswave.ll:
```diff
@@ -0,0 +1,37 @@
+; CHECK-NEXT: {{^;$}}
+
+; TODO: is there a way that we can split the input file into many?
+; as the shader flag will be true if any are set
```
The other option, that I think would be best, is to add this `RUN` line to each of the operation tests, eg in `WaveReadLaneAt.ll` in the parent directory.

inbelic commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -1091,4 +1092,4 @@ def CreateHandleFromBinding : DXILOp<217, createHandleFromBinding> {
   let result = HandleTy;
   let stages = [Stages<DXIL1_6, [all_stages]>];
   let attributes = [Attributes<DXIL1_0, [ReadNone]>];
-}
```
Artifact from github ui resolve conflicts. Will update with other review comments

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -30,6 +31,28 @@
 using namespace llvm;
 using namespace llvm::dxil;
 
+static dxil::Properties getOpCodeProperties(dxil::OpCode OpCode) {
+  dxil::Properties Props;
+  switch (OpCode) {
+#define DXIL_OP_PROPERTIES(OpCode, ...)                                        \
+  case OpCode:                                                                 \
+    Props = dxil::Properties{__VA_ARGS__};                                     \
+    break;
+#include "DXILOperation.inc"
+  }
+  return Props;
+}
+
+static bool checkWaveOps(Intrinsic::ID IID) {
+  switch (IID) {
+#define DXIL_OP_INTRINSIC(OpCode, IntrinsicID, ...)                            \
+  case IntrinsicID:                                                            \
+    return getOpCodeProperties(OpCode).IsWave;
+#include "DXILOperation.inc"
+  }
+  return false;
+}
```
There might be an argument for simply listing the 30/40 relevant intrinsic IDs in a switch statement here rather than having the whole infrastructure to map in reverse from the DXIL opcode to the directx intrinsic

llvm-beanz replied:
We are probably going to want to be able to map DXIL opcodes to intrinsic functions for the DXIL loading path though right? So it is infrastructure we need one way or another.

inbelic replied:
I think in either case, it is more a question if we want to continuing using the `IsWave` property (and `DXILProperty`s in general) in `DXIL.td` to model sub-groups of ops for shader flag analysis.

The removal of `DXILProperty` is outlined [here](https://github.com/llvm/llvm-project/issues/126298).

bob80905 commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -92,6 +115,8 @@ void ModuleShaderFlags::updateFunctionFlags(ComputedShaderFlags &CSF,
     // TODO: Set DX11_1_DoubleExtensions if I is a call to DXIL intrinsic
     // DXIL::Opcode::Fma https://github.com/llvm/llvm-project/issues/114554
+
+    CSF.WaveOps |= checkWaveOps(CI->getIntrinsicID());
```
There might be some weird behavior with the uninitialized value at 35 being returned if the switch statmeent on line 36 doesn't enter the case on line 38. checkWaveOps will return this uninitialized value's IsWave member on line 50. 
Or I might be reading this wrong. Is there a way for an uninitialized value to be returned?

inbelic replied:
Marking as resolved with the different appraoch

inbelic committed:
[DirectX][ShaderFlags] Add analysis for `WaveOps` flag

- add Shader Flag analysis for the `WaveOps` flag in DXILShaderFlags.cpp
- add testing of all currenlty supported wave ops in wave-ops.ll

inbelic committed:
clang formatting

bob80905 commented on llvm/test/CodeGen/DirectX/ShaderFlags/wave-ops.ll:
```diff
@@ -0,0 +1,79 @@
+; CHECK: ; Note: shader requires additional functionality:
+; CHECK-NEXT: ;       Wave level operations
+; CHECK-NEXT: ; Note: extra DXIL module flags:
+; CHECK-NEXT: {{^;$}}
```
Can you explain what this check does?

inbelic committed:
review: add test description

inbelic committed:
review: remove unneeded check

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -30,6 +30,40 @@
+  case Intrinsic::dx_wave_reduce_max:
+  case Intrinsic::dx_wave_reduce_umax:
+
+    // Currently unsupported intrinsics
+    // case Intrinsic::dx_WaveGetLaneCount:
+    // case Intrinsic::dx_WaveActiveAllEqual:
+    // case Intrinsic::dx_WaveActiveBallot:
+    // case Intrinsic::dx_WaveReadLaneFirst:
+    // case Intrinsic::dx_WaveActiveBit:
+    // case Intrinsic::dx_WavePrefixOp:
+    // case Intrinsic::dx_QuadReadLaneAt:
+    // case Intrinsic::dx_QuadOp:
+    // case Intrinsic::dx_WavePrefixBitCount:
+    // case Intrinsic::dx_WaveMatch:
+    // case Intrinsic::dx_WaveMultiPrefixOp:
+    // case Intrinsic::dx_WaveMultiPrefixBitCount:
+    // case Intrinsic::dx_QuadVote:
```
We have proposed spellings for all of these, so you could name these to match those if you like: https://github.com/llvm/wg-hlsl/blob/main/proposals/0014-consistent-naming-for-dx-intrinsics.md#wave-ops

Also clang-format does something kind of awkward here - might be better to put this comment outside of the switch.

bogner reviewed (approved):
LGTM!

inbelic committed:
[NFC] review: provide naming convention spellings and fix format
